pkg/apis/core/types.go

```diff
@@ -4255,11 +4255,44 @@ type NodeAddressType string
 
 // These are valid values of node address type
 const (
-	NodeHostName    NodeAddressType = "Hostname"
-	NodeExternalIP  NodeAddressType = "ExternalIP"
-	NodeInternalIP  NodeAddressType = "InternalIP"
-	NodeExternalDNS NodeAddressType = "ExternalDNS"
+	// NodeHostName identifies a name of the node. Although every node can be assumed
+	// to have a NodeAddress of this type, its exact syntax and semantics are not
+	// defined, and are not consistent between different clusters.
+	NodeHostName NodeAddressType = "Hostname"
+
+	// NodeInternalIP identifies an IP address which is assigned to one of the node's
+	// network interfaces. Every node should have at least one address of this type.
+	//
+	// An internal IP is normally expected to be reachable from every other node, but
+	// may not be visible to hosts outside the cluster. By default it is assumed that
+	// kube-apiserver can reach node internal IPs, though it is possible to configure
+	// clusters where this is not the case.
```

Comment:
I believe we're also asserting that any Pod can at least _unambigously identify_ an NodeInternalIP address as an internal IP, even if the Pod may be prevented from sending traffic to / from that internal Node IP.

Does that sound right?

For example, if I'm running as a Pod `foo` in some namespace in the cluster and I learn that 10.42.0.1 is a node IP address, that doesn't ever conflict with the IP address for a Service, another Pod, the API server, etc. I can recognise that a connection from 10.42.0.1 could only have originated from that node (maybe it is performing a TCP health check as I am executing on that particular node).

Reply:
>I can recognise that a connection from 10.42.0.1 could only have originated from that node (maybe it is performing a TCP health check as I am executing on that particular node).

I think this gets tricky given some edge scenarios regarding SNAT